account_proof.go

```diff
@@ -0,0 +1,64 @@
+package flow
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+type canonicalAcctProofWithoutTag struct {
+	Addr      []byte
+	Timestamp uint64
+}
+
+type canonicalAcctProofWithTag struct {
+	DomainTag []byte
+	Addr      []byte
+	Timestamp uint64
+}
+
+// NewAccountProofMsg creates a new account proof message for singing. The appDomainTag is optional and can be left
+// empty. Note that the resulting byte slice does not contain the user domain tag.
+func NewAccountProofMsg(addr string, timestamp int64, appDomainTag string) ([]byte, error) {
```

Comment:
This was implemented by reverse-engineering the [JS SDK implementation](https://github.com/onflow/fcl-js/blob/master/packages/fcl/src/wallet-utils/encode-provable-authn-signing.js).

Reply:
Can the `addr` param be typed as `Address`, or does it need to be a string?

Reply:
I'm not sure the logic in this function matches the FCL implementation, which is a bit hard to follow so I could be wrong. Did you test this against the JS version?

In this Go version:
- Address is passed to the RLP encoder as a byte array
- Domain tag is a byte array containing the UTF-8 bytes of a hexadecimal string
- Timestamp is a number

In the JS version:
- Address is passed as a hexadecimal string (with a 0x prefix)
- Domain tag is also passed as a hexadecimal string
- Timestamp is a number (I think)

Given that all of this is then RLP encoded, there's confusing encoding going on in both versions 🤔 ( hex -> rlp)

Reply:
I wondered while I was reverse engineering the JS SDK if the people who wrote the JS version intended for their message encoding to work the way it does. It's very strange.

Reply:
@bfbachmann After discussion with the FCL team, the encoding format was updated to follow this specification:

```js
MESSAGE = 
  USER_DOMAIN_TAG + // Array<Byte>, right padded to 32 bytes long
  RLP(
    APP_DOMAIN_TAG, // [Optional] Array<Byte>, right padded to 32 bytes long
    ADDRESS,  // Array<Byte>, left padded to 8 bytes long
    TIMESTAMP // Number
  )
```

The JS implementation has been updated in this PR: https://github.com/onflow/fcl-js/pull/1028

I would recommend that you replicate the test cases from JS implementation to test parity:

- These are the tests: https://github.com/onflow/fcl-js/blob/master/packages/fcl/src/wallet-utils/encode-provable-authn-signing.test.js
- The expected values are here: https://github.com/onflow/fcl-js/blob/master/packages/fcl/src/wallet-utils/__snapshots__/encode-provable-authn-signing.test.js.snap

You can get an equivalent test address like so:

```go
address := flow.HexToAddress("0xABC123DEF456")
```